### PR TITLE
Add tests for `wp_nonce_url()`.

### PR DESCRIPTION
**Pull request by costdev**
This PR revises the datasets from [PR 2201](https://github.com/WordPress/wordpress-develop/pull/2201) for accurate testing.

Trac ticket: https://core.trac.wordpress.org/ticket/54870

### CODE REVIEW

File: tests/phpunit/tests/functions/wpNonceUrl.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		);
+
+		$this->assertNotFalse(
+			wp_verify_nonce( $nonce, $action ),
+			'The nonce is invalid'
+		);
+	}
+
+	/**
+	 * Data provider
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/functions/wpNonceUrl.php]:
```suggestion
	 * Data provider for test_should_append_nonce_name_and_value().
```

File: tests/phpunit/tests/functions/wpNonceUrl.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+				'name'      => 'my_nonce',
+			),
+		);
+	}
+
+	/**
+	 * Tests that wp_nonce_url() handles existing query args.
+	 *
+	 * @dataProvider data_should_handle_existing_query_args
+	 * @ticket 54870
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/functions/wpNonceUrl.php]: Per [56070](https://core.trac.wordpress.org/ticket/56070), @ticket should precede @dataProvider and there should be a line break between them.

i.e.
```suggestion
	 * @ticket 54870
	 *
	 * @dataProvider data_should_handle_existing_query_args
```

File: tests/phpunit/tests/functions/wpNonceUrl.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	 *
+	 * @param string $actionurl URL to add nonce action.
+	 */
+	public function test_should_handle_existing_query_args( $actionurl ) {
+		$actionurl = preg_replace( '/&(?!amp;)/', '&amp;', $actionurl );
+		$this->assertStringContainsString( $actionurl, wp_nonce_url( $actionurl ) );
+	}
+
+	/**
+	 * Data provider.
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/functions/wpNonceUrl.php]:
```suggestion
	 * Data provider for test_should_handle_existing_query_args().
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Left some review

File: tests/phpunit/tests/functions/wpNonceUrl.php (hunk 1 above)
Comment (costdev) [reply tests/phpunit/tests/functions/wpNonceUrl.php]: Fixed in [5be5a43](https://github.com/WordPress/wordpress-develop/pull/2653/commits/5be5a4392bfffc703362ae3fde4abfbff63f04d2).

File: tests/phpunit/tests/functions/wpNonceUrl.php (hunk 2 above)
Comment (costdev) [reply tests/phpunit/tests/functions/wpNonceUrl.php]: Fixed in [5be5a43](https://github.com/WordPress/wordpress-develop/pull/2653/commits/5be5a4392bfffc703362ae3fde4abfbff63f04d2).

File: tests/phpunit/tests/functions/wpNonceUrl.php (hunk 3 above)
Comment (costdev) [reply tests/phpunit/tests/functions/wpNonceUrl.php]: Fixed in [5be5a43](https://github.com/WordPress/wordpress-develop/pull/2653/commits/5be5a4392bfffc703362ae3fde4abfbff63f04d2).

File: tests/phpunit/tests/functions/wpNonceUrl.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	 * Tests that wp_nonce_url() handles existing query args.
+	 *
+	 * @ticket 54870
+	 *
+	 * @dataProvider data_should_handle_existing_query_args
+	 *
+	 * @param string $actionurl URL to add nonce action.
+	 */
+	public function test_should_handle_existing_query_args( $actionurl ) {
+		$actionurl = preg_replace( '/&(?!amp;)/', '&amp;', $actionurl );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/functions/wpNonceUrl.php]: This looks a little like it is testing the code by running a copy-paste of the code. 

I think i'd prefer an `$expted` to be passed so it's clear what is been tested.

Comment (costdev) [reply tests/phpunit/tests/functions/wpNonceUrl.php]: Not quite a copy-paste. It's taking a realistic URL from the data provider, and ensuring that it has `&amp;`. In doing so, this verifies that `str_replace()` is called in `wp_nonce_url()` to convert `&amp;` to `&`.

I'm happy to change the variable on Line 108 from `$actionurl` to `$expected`. The parameter however is named `$actionurl` to increase familiarity between the source function and the test method/datasets for easier testing, so I'd like to keep the parameter name as-is. Does that work for you?

Comment (peterwilsoncc) [reply tests/phpunit/tests/functions/wpNonceUrl.php]: I was thinking more of using two params, `$actionulr, $expected`, to avoid surprises

Comment (costdev) [reply tests/phpunit/tests/functions/wpNonceUrl.php]: Hmm... Looking back over this test method and datasets, this PR needs more work to properly cover `wp_nonce_url`.

I'll ping you when I have this revision work done.

Comment (costdev) [reply tests/phpunit/tests/functions/wpNonceUrl.php]: @peterwilsoncc Ping!

File: tests/phpunit/tests/functions/wpNonceUrl.php [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	 *
+	 * @ticket 54870
+	 *
+	 * @dataProvider data_should_handle_existing_query_args
+	 *
+	 * @param string $actionurl URL to add nonce action.
+	 * @param string $expected  The expected result.
+	 */
+	public function test_should_handle_existing_query_args( $actionurl, $expected ) {
+		$this->assertStringContainsString( $expected, wp_nonce_url( $actionurl ) );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/functions/wpNonceUrl.php]:
```suggestion
		$this->assertStringStartsWith( $expected, wp_nonce_url( $actionurl ) );
```

Maybe assert that `strlen( expected ) + 10 === strlen( $url )` but no big deal.

**Review by peterwilsoncc — COMMENTED** [review]
LGTM, added a suggestion about a more precise assertion because I am the worst.

File: tests/phpunit/tests/functions/wpNonceUrl.php (hunk 5 above)
Comment (costdev) [reply tests/phpunit/tests/functions/wpNonceUrl.php]: - Assertion changed in [0f1a131](https://github.com/WordPress/wordpress-develop/pull/2653/commits/0f1a131b7d40858a7d3097e88d8957d3539ffef6).
- Assertion added in [a7d1441](https://github.com/WordPress/wordpress-develop/pull/2653/commits/a7d14417158ca36ca5ace20be11adfb26870709d).

**Review by peterwilsoncc — APPROVED** [review]
Dandy 🌻 

Thank you!

**Comment by peterwilsoncc** [comment]
Committed in https://core.trac.wordpress.org/changeset/54407 / https://github.com/WordPress/wordpress-develop/commit/426d20a60afbeefac2fba8e75dad97c0dcc2d42a

**Pull request closed without merging**